Add scan planning api request and response models, parsers

rahil-c opened the pull request:
Opening a new pr which just focues on the scan planning model and parsers based off the original pr here https://github.com/apache/iceberg/pull/11180/files#diff-eeec6df5a574c599e4bab34e926583f9e382ec55bf0519d8fbea382bdd25c6c5 

cc @amogh-jahagirdar @rdblue @danielcweeks @jackye1995 @nastra @singhpk234

rahil-c commented:
Thanks @amogh-jahagirdar for helping clean up the pr, it is greatly appreciated!

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/rest/responses/FetchPlanningResultResponse.java:
```diff
@@ -0,0 +1,133 @@
+      return this;
+    }
+
+    public Builder withSpecsById(Map<Integer, PartitionSpec> withSpecsById) {
```
I'm going to double check if we really need this or if it's correct to derive the specs by id from the file scan tasks themselves.

amogh-jahagirdar replied:
Verified it's correct to do without this but during serialization it'll require a double pass either over the delete files or the file scan tasks. I think doing a double pass over delete files in the average case is probably better it if we think that the average query will have far more data files than delete files. But of course the worst case is there is some multiple of delete files compared to data files.

It's nice to not have to pass this to the builder, but I also don't think it's worth overcomplicating since a server building the response can easily pass in the specs by ID.  I'll update if @rdblue @nastra @danielcweeks feel other wise.

Note, this does not actually get serialized it's just used during serialization itself when serializing the actual files involved.

rahil-c replied:
My understanding is we still need this `specsById` for ensuring the `date-file` and `delete-files` has the correct partition information during serialization time.

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/UnboundBaseFileScanTask.java:
```diff
@@ -0,0 +1,124 @@
+        .toString();
+  }
+
+  public FileScanTask bind(PartitionSpec spec, boolean caseSensitive) {
```
I think we should add tests for this method

amogh-jahagirdar replied:
My recent change to his branch should fix this, resolving

amogh-jahagirdar commented on api/src/main/java/org/apache/iceberg/exceptions/EntityNotFoundException.java:
```diff
@@ -0,0 +1,34 @@
+import com.google.errorprone.annotations.FormatMethod;
+
+/** Exception raised when an entity is not found. */
+public class EntityNotFoundException extends RESTException implements CleanableFailure {
```
Do we need this? I remember @rdblue mentioned a common exception for when a resource can't be found, but it doesn't seem like it's really being used anywhere

rahil-c replied:
I am not sure why we added this, I think we can remove it.

amogh-jahagirdar replied:
I'd say let's remove this for now since at the moment it's not being used, I think it's something we can add once there's clarity on how it would get used.

rahil-c replied:
Ack

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/rest/requests/PlanTableScanRequest.java:
```diff
@@ -0,0 +1,170 @@
+  private Long snapshotId;
+  private List<String> select;
+  private Expression filter;
+  private Boolean caseSensitive;
+  private Boolean useSnapshotSchema;
```
These should be primitive, fixing this

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/rest/requests/PlanTableScanRequestParser.java:
```diff
@@ -0,0 +1,142 @@
+      filter = ExpressionParser.fromJson(json.get(FILTER).textValue());
+    }
+
+    Boolean caseSensitive = true;
```
Should be primitive

nastra commented:
@rahil-c can you please make sure that my comments from https://github.com/apache/iceberg/pull/11180 are addressed here?

rahil-c commented:
> @rahil-c can you please make sure that my comments from #11180 are addressed here?

@nastra yes will try to address the comments from the original pr here.

cc @amogh-jahagirdar

amogh-jahagirdar reviewed (commented):
I think we're missing tests for the parsing of requests. 

Edit: NVM, we do have tests for those. The names of the tests were a bit different than I expected. Discussed with @rahil-c who'll make the names more consistent

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java:
```diff
@@ -511,7 +511,6 @@ public Table loadTable(SessionContext context, TableIdentifier identifier) {
             endpoints);
 
     trackFileIO(ops);
-
```
Can we undo this change, it's unnecessary

rahil-c replied:
Will fix this

rahil-c commented:
@nastra @amogh-jahagirdar Have addressed the comments from the main pr: https://github.com/apache/iceberg/pull/11180/files to this models only pr. If something looks off please let me know and i will do another pass as well.

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/ContentFileParser.java:
```diff
@@ -48,6 +48,97 @@ class ContentFileParser {
+        contentFile.specId());
+
+    generator.writeStartObject();
+    // ignore the ordinal position (ContentFile#pos) of the file in a manifest,
+    // as it isn't used and BaseFile constructor doesn't support it.
```
Don't think we need this comment, we can remove it

rahil-c replied:
@amogh-jahagirdar I actually am not the original author of that comment. 
<img width="898" alt="Screenshot 2024-11-27 at 11 15 47 AM" src="https://github.com/user-attachments/assets/b5219cee-3cbd-4128-9d96-d22bbf25df1c">

This was the pr that added it: https://github.com/apache/iceberg/commit/b8db3f00424c2cc2910a760b9068d38916e8c27a. Do you still want me to remove it?

rdblue replied:
I think it is okay since it is pre-existing. It's also good context since the position isn't in the spec and is used to help streaming readers keep incremental state.

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/rest/requests/FetchScanTasksRequest.java:
```diff
@@ -0,0 +1,67 @@
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.rest.RESTRequest;
+
+@SuppressWarnings("checkstyle:VisibilityModifier")
```
Do we need this?

rahil-c replied:
Let me see if I can get rid of this, this was for not having getter and setters for the the `plan-task` I believe.

rahil-c replied:
<img width="861" alt="Screenshot 2024-11-27 at 11 31 23 AM" src="https://github.com/user-attachments/assets/67e57d0f-8a63-44bd-819f-55e71ea082d3">


Im not sure if checkStyle has a bug or if im doing something incorrect. It seems that checkStyle is complaining that the private var `planTask` does not have an accessor method. However I do have the following in the class

```
public String planTask() {
    return planTask;
  }
```

And it seems to follow the pattern such as https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/rest/requests/UpdateTableRequest.java

If i remove the `@SuppressWarnings("checkstyle:VisibilityModifier")` it will continue to fail the build unfortunately so will keep this for now.

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/rest/requests/PlanTableScanRequestParser.java:
```diff
@@ -0,0 +1,126 @@
+    Preconditions.checkArgument(null != request, "Invalid plan table scan request: null");
+
+    gen.writeStartObject();
+    if (request.snapshotId() != null) {
+      gen.writeNumberField(SNAPSHOT_ID, request.snapshotId());
+    }
+
+    if (request.startSnapshotId() != null) {
+      gen.writeNumberField(START_SNAPSHOT_ID, request.startSnapshotId());
+    }
+
+    if (request.endSnapshotId() != null) {
+      gen.writeNumberField(END_SNAPSHOT_ID, request.endSnapshotId());
+    }
```
On second look, the way this is written is a bit confusing after reading again since on the surface it seems like it's possible serializing snapshotId/startingSnapshotId/endSnapshotID even though it's either a point in time scan or an incremental scan.  To be clear, I know that validate checks this, I'm more so talking about how a reader would interpret this code.

concretely I'd recommend replacing this with:

````
private void serializeSnapshotIdForScan(JsonGenerator gen, request) {
    if (request.snapshotId() != null) {
         gen.writeNumberField(SNAPSHOT_ID, request.snapshotId());
     } else {
        gen.writeNumberField(START_SNAPSHOT_ID, request.startSnapshotId());
        gen.writeNumberField(END_SNAPSHOT_ID, request.endSnapshotId());
     }
}
````

I think then you'd also be able to get rid of cyclomatic complexity override above.

rahil-c replied:
Will follow this suggestion thanks @amogh-jahagirdar

rahil-c replied:
Actually when looking at this again are we sure this is what we want? When reading the spec again https://github.com/apache/iceberg/blob/main/open-api/rest-catalog-open-api.yaml#L609 
```
  If the request does not include either incremental or
        point-in-time config properties, scan planning should produce a
        point-in-time scan of the latest snapshot in the table's main branch
```

It seems possible that the client does not need to pass either a snapshotId, or start and end snapshot ids, and that the server can just plan with the latest snapshot id if nothing was provided. I would assume then at serialize time this would be an issue with the following code, 

```
private void serializeSnapshotIdForScan(JsonGenerator gen, request) {
    if (request.snapshotId() != null) {
         gen.writeNumberField(SNAPSHOT_ID, request.snapshotId());
     } else {
        gen.writeNumberField(START_SNAPSHOT_ID, request.startSnapshotId());
        gen.writeNumberField(END_SNAPSHOT_ID, request.endSnapshotId());
     }
}

```
as it may hit an error if the start and end ids were also null. I think the way it currently works because we the null check around each of the ids. Let me know what you think though

amogh-jahagirdar commented on core/src/main/java/org/apache/iceberg/RESTFileScanTaskParser.java:
```diff
@@ -0,0 +1,89 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.util.JsonUtil;
+
+public class RESTFileScanTaskParser {
```
On second look, shouldn't this be in the REST package?

rahil-c replied:
If we move this to the REST package then we run into issues that we saw in the past with the Data and Delete Files being package private.

<img width="425" alt="Screenshot 2024-11-27 at 11 53 09 AM" src="https://github.com/user-attachments/assets/737ea1eb-a60b-4808-be62-c00b5031d3e1">

I think we can keep where it is currently, so we can avoid making anything public.

amogh-jahagirdar replied:
Thanks for the reminder, I think in my PR to your branch, I forgot to address this part I think it's worth seeing how we can refactor this so that we can actually move `RESTFileScanTaskParser` in the REST package...it kind of stands out by not really following existing convention

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/responses/TestFetchScanTasksResponseParser.java:
```diff
@@ -0,0 +1,180 @@
+            + "}";
+
+    FetchScanTasksResponse fromResponse = FetchScanTasksResponseParser.fromJson(json);
+    // Need to make a new response with partitionSpec set
```
I'd say remove the comment

rahil-c replied:
will remove it

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/responses/TestFetchScanTasksResponseParser.java:
```diff
@@ -0,0 +1,180 @@
+    String json = FetchScanTasksResponseParser.toJson(response, false);
+    assertThat(json).isEqualTo(expectedToJson);
+
+    // make an unbound json where you expect to not have partitions for the data file,
+    // delete files as service does not send parition spec
```
Generally try and avoid the second person in comments, "create a response where the file scan tasks are unbound" . It's not entirely clear or accurate to say "not have partitions for the data file", it's more so that there's no partitions that are in the deserialized version until it's bound.

rahil-c replied:
Will fix this

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/responses/TestFetchScanTasksResponseParser.java:
```diff
@@ -0,0 +1,180 @@
+
+    FetchScanTasksResponse fromResponse = FetchScanTasksResponseParser.fromJson(json);
+    // Need to make a new response with partitionSpec set
+    FetchScanTasksResponse copyResponse =
```
The name `copyResponse` doesn't really indicate much imo, I'd say `responseWithSpecs`

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/responses/TestFetchScanTasksResponseParser.java:
```diff
@@ -0,0 +1,180 @@
+        FetchScanTasksResponse.builder()
+            .withFileScanTasks(List.of(fileScanTask))
+            .withDeleteFiles(List.of(FILE_A_DELETES))
+            // assume you have set this already
```
Same as below, we generally don't use second person in comments, let's remove/reword instances of those

rahil-c replied:
will remove

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/responses/TestFetchScanTasksResponseParser.java:
```diff
@@ -0,0 +1,180 @@
+            + "\"residual-filter\":{\"type\":\"eq\",\"term\":\"id\",\"value\":1}}]"
+            + "}";
+
+    FetchScanTasksResponse fromResponse = FetchScanTasksResponseParser.fromJson(json);
```
`response` should suffice for the purpose of this test?  `fromResponse` doesn't really indicate too much imo,

rahil-c replied:
I used the naming `fromResponse` since the parser is invoking `fromJson`. Maybe we can make this as `nit` for now?

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/responses/TestFetchScanTasksResponseParser.java:
```diff
@@ -0,0 +1,180 @@
+
+    // can't do an equality comparison on PlanTableScanRequest because we don't implement
+    // equals/hashcode
+    assertThat(FetchScanTasksResponseParser.toJson(copyResponse, false))
+        .isEqualTo(expectedFromJson);
```
Actually on second look, I feel like this test reads a bit confusing. Shouldn't this be asserting that the deserialized JSON from line 147 is equal to the unbound JSON on line 152? What are we testing by building a copy with the specs? We already test that we serialize correctly, above so it seems duplicative to test that again.

rahil-c replied:
Actually I think these are asserting two different things.

The first set of asserts, show that serialization of tasks works correctly when there is a specsByID set, however the expectation is that `partition` data should be present.

```
FetchScanTasksResponse response =
        FetchScanTasksResponse.builder()
            .withFileScanTasks(List.of(fileScanTask))
            .withDeleteFiles(List.of(FILE_A_DELETES))
            .withSpecsById(PARTITION_SPECS_BY_ID)
            .build();

    String expectedToJson =
        "{"
            + "\"delete-files\":[{\"spec-id\":0,\"content\":\"POSITION_DELETES\","
            + "\"file-path\":\"/path/to/data-a-deletes.parquet\",\"file-format\":\"PARQUET\","
            + "\"partition\":{\"1000\":0},\"file-size-in-bytes\":10,\"record-count\":1}],"
            + "\"file-scan-tasks\":["
            + "{\"data-file\":{\"spec-id\":0,\"content\":\"DATA\",\"file-path\":\"/path/to/data-a.parquet\","
            + "\"file-format\":\"PARQUET\",\"partition\":{\"1000\":0},"
            + "\"file-size-in-bytes\":10,\"record-count\":1,\"sort-order-id\":0},"
            + "\"delete-file-references\":[0],"
            + "\"residual-filter\":{\"type\":\"eq\",\"term\":\"id\",\"value\":1}}]"
            + "}";

    String json = FetchScanTasksResponseParser.toJson(response, false);
    assertThat(json).isEqualTo(expectedToJson);
```


The second set of asserts, show that deserialization of tasks which will have unbound tasks works correctly, and that `partition` data will be empty in this case.

```
// create a response where the file scan tasks are unbound
    String expectedFromJson =
        "{"
            + "\"delete-files\":[{\"spec-id\":0,\"content\":\"POSITION_DELETES\","
            + "\"file-path\":\"/path/to/data-a-deletes.parquet\",\"file-format\":\"PARQUET\","
            + "\"partition\":{},\"file-size-in-bytes\":10,\"record-count\":1}],"
            + "\"file-scan-tasks\":["
            + "{\"data-file\":{\"spec-id\":0,\"content\":\"DATA\",\"file-path\":\"/path/to/data-a.parquet\","
            + "\"file-format\":\"PARQUET\",\"partition\":{},"
            + "\"file-size-in-bytes\":10,\"record-count\":1,\"sort-order-id\":0},"
            + "\"delete-file-references\":[0],"
            + "\"residual-filter\":{\"type\":\"eq\",\"term\":\"id\",\"value\":1}}]"
            + "}";

    FetchScanTasksResponse deserializedResponse = FetchScanTasksResponseParser.fromJson(json);
    FetchScanTasksResponse responseWithSpecs =
        FetchScanTasksResponse.builder()
            .withPlanTasks(deserializedResponse.planTasks())
            .withDeleteFiles(deserializedResponse.deleteFiles())
            .withFileScanTasks(deserializedResponse.fileScanTasks())
            .withSpecsById(PARTITION_SPECS_BY_ID)
            .build();

    assertThat(FetchScanTasksResponseParser.toJson(responseWithSpecs, false))
        .isEqualTo(expectedFromJson);
```

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/responses/TestFetchScanTasksResponseParser.java:
```diff
@@ -0,0 +1,180 @@
+  }
+
+  @Test
+  public void roundTripSerdeWithFileScanTasks() {
```
I know `roundTrip` prefix was probably copied from existing tests, but this protocol doesn't really round trip like other JSON messages due to the bound/unbound specs involved. Maybe just call this `serdeWithFileScanTasks`

rahil-c replied:
Yea this is a good point will remove the `roundTrip` naming from all the response parser tests

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/responses/TestPlanTableScanResponseParser.java:
```diff
@@ -0,0 +1,244 @@
+  }
+
+  @Test
+  public void roundTripSerdeWithInvalidPlanStatusSubmittedWithDeleteFilesNoFileScanTasksPresent() {
```
I don't think any of the failure case tests should have the prefix "round trip", nothing is really roundtripping here

rahil-c replied:
Will fix this

amogh-jahagirdar commented on core/src/test/java/org/apache/iceberg/rest/requests/TestPlanTableScanRequestParser.java:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  @Test
+  public void roundTripSerdeWithEmptyRequestAndDefaultsPresent() {
```
Same as below, let's double check that we're not unnecessarily prefixing the failure case tests with `round trip` since nothing is really round tripping in case of failures

rahil-c replied:
will fix this

rahil-c commented:
@amogh-jahagirdar Was wondering if there is another issues, or if you think we can land this?

rahil-c commented:
Reviving this pr.

rdblue commented on core/src/main/java/org/apache/iceberg/ContentFileParser.java:
```diff
@@ -48,6 +48,97 @@ class ContentFileParser {
+      ContentFile<?> contentFile, PartitionSpec spec, JsonGenerator generator) throws IOException {
+    Preconditions.checkArgument(contentFile != null, "Invalid content file: null");
+    Preconditions.checkArgument(spec != null, "Invalid partition spec: null");
+    Preconditions.checkArgument(generator != null, "Invalid JSON generator: null");
```
We generally try to avoid exposing `JsonGenerator` or other Jackson classes in public methods. Can this be package-private? See [`SnapshotParser` for an example](https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/SnapshotParser.java#L55) of how other parsers handle visibility.

rdblue commented on core/src/main/java/org/apache/iceberg/ContentFileParser.java:
```diff
@@ -27,7 +27,7 @@
 import org.apache.iceberg.util.JsonUtil;
 
-class ContentFileParser {
+public class ContentFileParser {
```
Why is it necessary to make this constructor public?

rahil-c replied:
@rdblue @amogh-jahagirdar 

Currently the `ContentFileParser` is located within package `org.apache.iceberg` as opposed to something like `org.apache.iceberg.rest.responses`. The reason this parser is located within this package is that it is able to access private classes such as `ContentFile`, `GenericDataFile` etc.

However it seems we have placed the `TableScanResponseParser` which is what is used by the majority of endpoints in `org.apache.iceberg.rest.responses`, and this invokes the `ContentFileParser`. When i tried removing the `public` from content file parser naturally got issues.

rdblue commented on core/src/main/java/org/apache/iceberg/ContentFileParser.java:
```diff
@@ -48,6 +48,97 @@ class ContentFileParser {
+    generator.writeEndObject();
+  }
+
+  public static ContentFile<?> unboundContentFileFromJson(JsonNode jsonNode) {
```
Can this be package-private?

rdblue commented on core/src/main/java/org/apache/iceberg/ContentFileParser.java:
```diff
@@ -48,6 +48,97 @@ class ContentFileParser {
+
+    long fileSizeInBytes = JsonUtil.getLong(FILE_SIZE, jsonNode);
+    Metrics metrics = metricsFromJson(jsonNode);
+    ByteBuffer keyMetadata = JsonUtil.getByteBufferOrNull(KEY_METADATA, jsonNode);
```
This doesn't match the write side, which uses `SingleValueParser`. Can you also use `SingleValueParser` here? Or use `JsonUtil` above?

rdblue commented on core/src/main/java/org/apache/iceberg/UnboundBaseFileScanTask.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    String schemaString = SchemaParser.toJson(spec.schema());
+    String specString = PartitionSpecParser.toJson(spec);
```
I don't think it is a good idea to convert the schema and spec to JSON in every task.

amogh-jahagirdar replied:
That is a good point, this would be on the hot path so we may need to pass in the schemaString and specString to the task from the outside rather than converting it on every bind

rdblue commented on core/src/main/java/org/apache/iceberg/UnboundBaseFileScanTask.java:
```diff
@@ -0,0 +1,63 @@
+
+    String schemaString = SchemaParser.toJson(spec.schema());
+    String specString = PartitionSpecParser.toJson(spec);
+    ResidualEvaluator boundResidual = ResidualEvaluator.of(spec, filter, caseSensitive);
```
What if filter is null?

rahil-c commented:
@amogh-jahagirdar @singhpk234
